src/lib/optimizer/strategy/predicate_placement_rule.cpp

```diff
@@ -41,19 +41,54 @@ void PredicatePlacementRule::_push_down_traversal(const std::shared_ptr<Abstract
                                                   std::vector<std::shared_ptr<AbstractLQPNode>>& push_down_nodes,
                                                   AbstractCardinalityEstimator& estimator) {
   const auto input_node = current_node->input(input_side);
-  if (!input_node) return;  // Allow calling without checks
+  if (!input_node) {
+    Assert(push_down_nodes.empty(), "Expected pushdown nodes to be already inserted.");
+    return;  // Allow calling without checks
+  }
 
-  // A helper method for cases where the input_node does not allow us to proceed
+  // A helper method for cases where the input_node does not allow us to proceed.
   const auto handle_barrier = [&]() {
     _insert_nodes(current_node, input_side, push_down_nodes);
 
-    if (input_node->left_input()) {
+    const auto& barrier_node = input_node;
+    const auto barrier_node_is_pushdown_predicate = [&barrier_node]() {
+      if (barrier_node->type == LQPNodeType::Predicate) {
+        const auto predicate_node = std::static_pointer_cast<PredicateNode>(barrier_node);
+        return !_is_expensive_predicate(predicate_node->predicate());
+      } else if (barrier_node->type == LQPNodeType::Join) {
+        const auto join_mode = std::static_pointer_cast<JoinNode>(barrier_node)->join_mode;
+        return join_mode == JoinMode::Semi || join_mode == JoinMode::AntiNullAsTrue ||
+               join_mode == JoinMode::AntiNullAsFalse;
+      }
+      return false;
+    }();
+
+    auto next_push_down_traversal_root = barrier_node;
+    if (barrier_node_is_pushdown_predicate) {
+      // The barrier node is a predicate eligible for pushdown. Therefore, we would like it to be covered by the next
+      // recursion of _push_down_traversal. However, since _push_down_traversal looks at the inputs of a given
+      // root node only, barrier_node would not be pushed down.
+      // To allow for a pushdown of barrier_node, a temporary root node is inserted above it. Afterwards, the temporary
+      // root node is used as a root for the next _push_down_traversal recursion. As a result, barrier_node will be seen
+      // as an input of the temporary root node, and thus be pushed down by the rule, if possible.
```

Comment:
Formatting: try to fully use the 120 characters of a line before inserting a line break